scripts/common/timer.ts

```diff
@@ -88,43 +96,321 @@ export interface RunSubsegment {
 	stats: RunStats;
 }
 
-export function getSplitSegmentTime(runSplits: RunSplits, segmentIndex: number, subsegmentIndex: number): number {
-	if (subsegmentIndex > 0) {
-		return (
-			runSplits.segments[segmentIndex].subsegments[subsegmentIndex].timeReached -
-			runSplits.segments[segmentIndex].subsegments[subsegmentIndex - 1].timeReached
-		);
+export interface RunStats {
+	maxOverallSpeed: float;
+	maxHorizontalSpeed: float;
+
+	overallDistanceTravelled: float;
+	horizontalDistanceTravelled: float;
+
+	jumps: uint16;
+	strafes: uint16;
+}
+
+export type RunStatType = keyof RunStats;
+
+export const RunStatsProperties: Record<RunStatType, { unit: string; isPositiveGood: boolean }> = {
+	maxOverallSpeed: {
+		unit: 'ups',
+		isPositiveGood: true
+	},
+	maxHorizontalSpeed: {
+		unit: 'ups',
+		isPositiveGood: true
+	},
+	overallDistanceTravelled: {
+		unit: 'units',
+		isPositiveGood: false
+	},
+	horizontalDistanceTravelled: {
+		unit: 'units',
+		isPositiveGood: false
+	},
+	jumps: {
+		unit: 'jumps',
+		isPositiveGood: false
+	},
+	strafes: {
+		unit: 'strafes',
+		isPositiveGood: false
 	}
+};
 
-	if (segmentIndex > 0) {
-		return (
-			runSplits.segments[segmentIndex].subsegments[subsegmentIndex].timeReached -
-			runSplits.segments[segmentIndex - 1].subsegments.at(-1).timeReached
-		);
+//#endregion
+//#region Utility Functions
+
+export function findSubsegmentComparison(
+	baseSplits: RunSplits,
+	compSplits: RunSplits | null,
+	majorNum: number,
+	minorNum: number
+): [RunSubsegment, RunSubsegment | null] {
+	const baseSeg = baseSplits.segments[majorNum - 1];
+	const baseSubsegIdx = baseSeg.subsegments.findIndex((ss) => ss.minorNum === minorNum);
+
+	// If this happens, something's gone wrong in the caller logic.
+	if (baseSubsegIdx === -1) {
+		throw new Error(`Missing subsegment for majorNum: ${majorNum}, minorNum: ${minorNum}`);
 	}
 
-	return 0;
+	if (!compSplits) {
+		return [baseSeg.subsegments[baseSubsegIdx], null];
+	}
+
+	const compSeg = compSplits.segments[majorNum - 1];
+	const compSubsegIdx = compSeg.subsegments.findIndex((ss) => ss.minorNum === minorNum);
+
+	// Easy case, base and comp are at same position in splits: return respective subsegs
+	//   Base ----S1---Cp1-----Cp2----
+	//   Comp ---S1----Cp1----Cp2-----
```

Comment:
(nit) The hyphens don't line up exactly

Reply:
That's deliberate, to signify the CPs being hit at different points in time.